app/frontend/wallet/account-manager.ts

```diff
@@ -46,29 +49,66 @@ const AccountManager = ({
   }
 
   async function discoverAccounts() {
+    if (
+      maxAccountIndex === accounts.length - 1 || accounts.length > 0
+        ? !(await accounts[accounts.length - 1].isAccountUsed())
+        : false
+    ) {
+      return accounts
+    }
+
     const isBulkExportSupported = cryptoProvider.isFeatureSupported(
       CryptoProviderFeature.BULK_EXPORT
     )
     const shouldExplore =
       config.shouldExportPubKeyBulk && config.isShelleyCompatible && isBulkExportSupported
-    async function _discoverNextAccount(accountIndex: number) {
-      const newAccount = accounts[accountIndex] || discoverNextAccount()
-      const isAccountUsed = await newAccount.isAccountUsed()
-      if (accountIndex === accounts.length) await addNextAccount(newAccount)
-
-      return (
-        shouldExplore &&
-        isAccountUsed &&
-        accountIndex < maxAccountIndex &&
-        (await _discoverNextAccount(accountIndex + 1))
+
+    if (!shouldExplore) {
+      await exploreNextAccount()
+      return accounts
+    }
+
+    for (const [accountIndexStart, accountIndexEnd] of makeBulkAccountIndexIterator()) {
+      const _accountIndexStart = Math.max(accountIndexStart, accounts.length)
+      const _accountIndexEnd = Math.min(accountIndexEnd, maxAccountIndex)
+
+      if (_accountIndexStart > _accountIndexEnd) {
```

Comment:
how can this condition ever evaluate to true? My first impression would be that this could happen only if the data were somehow malformed - what am I missing here?

Reply:
I see, when trimming the start and end by the batches this can indeed happen